src/Host/Client/Impl/Host/RemoteCredentialsDecorator.cs

```diff
@@ -17,21 +17,26 @@ internal class RemoteCredentialsDecorator : ICredentialsDecorator {
         private readonly ISecurityService _securityService;
         private readonly IMainThread _mainThread;
         private Credentials _credentials;
+        private object _credentialsLock;
         private readonly string _authority;
         private readonly AsyncReaderWriterLock _lock;
-        private bool _credentialsAreValid;
         private readonly string _workspaceName;
 
         public RemoteCredentialsDecorator(string credentialAuthority, string workspaceName, ISecurityService securityService, IMainThread mainThread) {
             _securityService = securityService;
             _mainThread = mainThread;
             _authority = credentialAuthority;
             _lock = new AsyncReaderWriterLock();
-            _credentialsAreValid = true;
+            _credentialsLock = new object();
```

Comment:
There is already lock in this class - `_lock`. Since you need both sync and async locking, you can use `SemaphoreSlim` instead of `AsyncReaderWriterLock`